src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -651,110 +530,15 @@ public static string FormatSingle(float value, string? format, NumberFormatInfo
-            }
-
-            number.CheckConsistency();
-
-            // When the number is known to be roundtrippable (either because we requested it be, or
-            // because we know we have enough digits to satisfy roundtrippability), we should validate
-            // that the number actually roundtrips back to the original result.
-
-            Debug.Assert(((precision != -1) && (precision < SinglePrecision)) || (BitConverter.SingleToInt32Bits(value) == BitConverter.SingleToInt32Bits(NumberToFloat<float>(ref number))));
-
-            if (fmt != 0)
-            {
-                if (precision == -1)
-                {
-                    Debug.Assert((fmt == 'G') || (fmt == 'g') || (fmt == 'R') || (fmt == 'r'));
-
-                    // For the roundtrip and general format specifiers, when returning the shortest roundtrippable
-                    // string, we need to update the maximum number of digits to be the greater of number.DigitsCount
-                    // or SinglePrecision. This ensures that we continue returning "pretty" strings for values with
-                    // less digits. One example this fixes is "-60", which would otherwise be formatted as "-6E+01"
-                    // since DigitsCount would be 1 and the formatter would almost immediately switch to scientific notation.
-
-                    nMaxDigits = Math.Max(number.DigitsCount, SinglePrecision);
-                }
-                NumberToString(ref vlb, ref number, fmt, nMaxDigits, info);
-            }
-            else
-            {
-                Debug.Assert(precision == SinglePrecisionCustomFormat);
-                NumberToStringFormat(ref vlb, ref number, format, info);
-            }
-            return null;
-        }
-
-        public static string FormatHalf(Half value, string? format, NumberFormatInfo info)
-        {
-            var vlb = new ValueListBuilder<char>(stackalloc char[CharStackBufferSize]);
-            string result = FormatHalf(ref vlb, value, format, info) ?? vlb.AsSpan().ToString();
-            vlb.Dispose();
-            return result;
-        }
-
-        /// <summary>Formats the specified value according to the specified format and info.</summary>
-        /// <returns>
-        /// Non-null if an existing string can be returned, in which case the builder will be unmodified.
-        /// Null if no existing string was returned, in which case the formatted output is in the builder.
-        /// </returns>
-        private static unsafe string? FormatHalf<TChar>(ref ValueListBuilder<TChar> vlb, Half value, ReadOnlySpan<char> format, NumberFormatInfo info) where TChar : unmanaged, IUtfChar<TChar>
+        private static unsafe string? FormatFloat<TNumber, TChar>(ref ValueListBuilder<TChar> vlb, TNumber value, ReadOnlySpan<char> format, NumberFormatInfo info)
+            where TNumber : unmanaged, IBinaryFloatParseAndFormatInfo<TNumber>
+            where TChar : unmanaged, IUtfChar<TChar>
```

Comment:
A simple `double.ToString` would fail at the following assert of this method (GitHub doesn't let me comment on it):
`typeof(TChar) == typeof(char) || typeof(TChar) == typeof(byte)`

@dotnet/jit-contrib Can you help on this for how this assert can fail? Can it be a JIT issue?

Reply:
@huoyaoyuan are you sure it's that assert? it looks like it's actually
```cs
Debug.Assert(((precision != -1) && (precision < TNumber.MaxRoundTripDigits)) || (TNumber.FloatToBits(value) == TNumber.FloatToBits(NumberToFloat<TNumber>(ref number))));
```

Reply:
I see in this log: https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-102683-merge-12f787eac7714e67aa/Microsoft.Bcl.Numerics.Tests/1/console.8b35101e.log?helixlogtype=result

```
   at System.Number.FormatFloat[TNumber,TChar](ValueListBuilder`1& vlb, TNumber value, ReadOnlySpan`1 format, NumberFormatInfo info) in /_/src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:line 537
   at System.Number.TryFormatFloat[TNumber,TChar](TNumber value, ReadOnlySpan`1 format, NumberFormatInfo info, Span`1 destination, Int32& charsWritten) in /_/src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:line 517
   at System.Text.ValueStringBuilder.AppendFormatHelper(IFormatProvider provider, String format, ReadOnlySpan`1 args) in /_/src/libraries/System.Private.CoreLib/src/System/Text/ValueStringBuilder.AppendFormat.cs:line 214
   at System.String.FormatHelper(IFormatProvider provider, String format, ReadOnlySpan`1 args) in /_/src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs:line 528
```

Reply:
Does it still repro with DOTNET_ReadyToRun=0 and DOTNET_JitMinOpts=0 ? Also, what's TChar if it's not byte or char?

Reply:
When I tried to print something locally, `TChar` was `char`. Will test with R2R/opt settings tomorrow.

Reply:
It still reproduces when togglling R2R and MinOpts. A simple `123.456.ToString` can trigger it with corerun. I'll play with JITDump.

Reply:
`****** DONE compiling System.Number:FormatFloat[double,ushort]`
It becomes `ushort` somewhere

Reply:
@EgorBo It does look like a JIT issue.
The issue is in this method:
```csharp
        public static string FormatFloat<TNumber>(TNumber value, string? format, NumberFormatInfo info)
            where TNumber : unmanaged, IBinaryFloatParseAndFormatInfo<TNumber>
        {
            var vlb = new ValueListBuilder<char>(stackalloc char[CharStackBufferSize]);
            string result = FormatFloat(ref vlb, value, format, info) ?? vlb.AsSpan().ToString();
            vlb.Dispose();
            return result;
        }
```
I've confirmed with ILSpy that IL uses `char`:
```il
IL_0008: newobj instance void valuetype System.Span`1<char>::.ctor(void*, int32)
IL_001a: call valuetype System.ReadOnlySpan`1<char> System.String::op_Implicit(string)
IL_0020: call string System.Number::FormatFloat<!!TNumber, char>(valuetype System.Collections.Generic.ValueListBuilder`1<!!1>&, !!0, valuetype System.ReadOnlySpan`1<char>, class System.Globalization.NumberFormatInfo)
```

But in JITDump it says `ushort` is used:
```

STMT00008 ( ??? ... ??? )
               [000018] S-C-G------                         *  CALL      void   System.String:op_Implicit(System.String):System.ReadOnlySpan`1[ushort]
               [000024] ----------- retbuf                  +--*  LCL_ADDR  byref  V12 tmp4         [+0]
               [000017] ----------- arg1                    \--*  LCL_VAR   ref    V01 arg1         
Marked V12 as a single def temp

    [ 3]  31 (0x01f) ldarg.2
    [ 4]  32 (0x020) call 2B0003C4
In Compiler::impImportCall: opcode is call, kind=0, callRetType is ref, structSize is 0
Calling impNormStructVal on:
               [000025] -----------                         *  LCL_VAR   struct<System.ReadOnlySpan`1, 16> V12 tmp4         
resulting tree:
               [000025] -----------                         *  LCL_VAR   struct<System.ReadOnlySpan`1, 16> V12 tmp4         

lvaGrabTemp returning 13 (V13 tmp5) called for impSpillStackEnsure.


STMT00009 ( 0x01F[--] ... ??? )
               [000028] DAC-G------                         *  STORE_LCL_VAR ref    V13 tmp5         
               [000027] --C-G------                         \--*  CALL      ref    System.Number:FormatFloat[double,ushort](byref,double,System.ReadOnlySpan`1[ushort],System.Globalization.NumberFormatInfo):System.String
               [000020] ----------- arg0                       +--*  LCL_VAR   byref  V10 tmp2         
               [000022] ----------- arg1                       +--*  LCL_VAR   double V11 tmp3         
               [000025] ----------- arg2                       +--*  LCL_VAR   struct<System.ReadOnlySpan`1, 16> V12 tmp4         
               [000026] ----------- arg3                       \--*  LCL_VAR   ref    V02 arg2         
Marked V13 as a single def temp
```

Full JITDump is here: 
[JITDUMP.txt](https://github.com/dotnet/runtime/files/15457737/JITDUMP.txt)

I'm using Debug build of CLR.

Reply:
> But in JITDump it says ushort is used:

Afair, JIT renders `char` as `ushort` everywhere so it doesn't look to be the problem. Did you confirm that the assert fails because `TChar` is actually `ushort` ?

Reply:
When I tried your branch few days ago, it failed on the other assert rather than that

Reply:
Is JIT lowering `char` to `ushort` in some way? `Console.WriteLine(typeof(TChar).ToString())` prints `System.Char`.

Reply:
Like I said - it's how JIT renders it internally (after e.g.
```
static void Main()
{
    Foo<char>('1');
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void Foo<T>(T t) { }
```
JitDump, JitDisasm and JitDisasmSummary will show the compilation as:
```
Proga:Foo[ushort](ushort)
```

Reply:
Yes the assert does come from the line you pointed. Sorry for the noise. I'm trying to debug.

Reply:
![image](https://github.com/dotnet/runtime/assets/523221/35bef90e-68cf-4400-9528-70c9e0077902)

I've just tried locally again and it can confirm it points to that assert